src/toil/provisioners/clusterScaler.py

```diff
@@ -84,150 +85,180 @@ def get(self):
+    calculate an approximate minimum number of nodes that will fit the
+    given list of jobs. To run the bin-packing algorithm, use the
+    binPack method on a list of job shapes, and use getRequiredNodes
+    to get how many nodes it thinks are required.
+
+    :param list nodeShapes: A list of possible types of nodes that can be launched (as "Shape"s).
+    :param targetTime: The time before which all jobs should at least be started.
     """
-    logger.debug('Running bin packing for node shapes %s and %s job(s).', nodeShapes, len(jobShapes))
-    # Sort in descending order from largest to smallest. The FFD like-strategy will pack the jobs in order from longest
-    # to shortest.
-    jobShapes.sort()
-    jobShapes.reverse()
-
-
-    #Prioritize preemptable node shapes with the lowest memory
-    nodeShapes.sort(key=lambda nS: not nS.preemptable)
-    nodeShapes.sort(key=lambda nS: nS.memory)
-    
-    assert len(jobShapes) == 0 or jobShapes[0] >= jobShapes[-1]
+    def __init__(self, nodeShapes, targetTime=3600):
+        self.nodeShapes = nodeShapes
+        self.targetTime = targetTime
+        # Prioritize preemptable node shapes with the lowest memory
+        self.nodeShapes.sort(key=lambda nS: not nS.preemptable)
+        self.nodeShapes.sort(key=lambda nS: nS.memory)
+        self.nodeReservations = {nodeShape:[] for nodeShape in nodeShapes}  # The list of node reservations
+
+    def binPack(self, jobShapes):
+        """Pack a list of jobShapes into the fewest nodes reasonable. Can be run multiple times."""
+        logger.debug('Running bin packing for node shapes %s and %s job(s).', self.nodeShapes, len(jobShapes))
+        # Sort in descending order from largest to smallest. The FFD like-strategy will pack the jobs in order from longest
+        # to shortest.
+        jobShapes.sort()
+        jobShapes.reverse()
+        assert len(jobShapes) == 0 or jobShapes[0] >= jobShapes[-1]
+        for jS in jobShapes:
+            self.addJobShape(jS)
+
+    def addJobShape(self, jobShape):
+        """
+        Function adds the job to the first node reservation in which it will fit (this
+        is the bin-packing aspect)
+        """
+        for nodeShape, nodeReservations in iteritems(self.nodeReservations):
+            for nodeReservation in nodeReservations:
+                if nodeReservation.attemptToAddJob(jobShape, nodeShape, self.targetTime):
+                    # We succeeded adding the job to this node reservation. Now we're done.
+                    return
+
+        # If we get here, the job didn't fit into any existing node
+        # reservations. Assign to the smallest node shape that will
+        # fit this job, prioritizing preemptable nodes
+        consideredNodes = [nodeShape for nodeShape in self.nodeShapes if nodeShape.cores >= jobShape.cores and nodeShape.memory >= jobShape.memory and nodeShape.disk >= jobShape.disk and (jobShape.preemptable or not nodeShape.preemptable)]
```

Comment:
This is essentially duplicating the NodeReservation.fits() function. Not sure if it's worth the effort to avoid duplicated code, but maybe NodeReservation should be a subclass of Shape, and Shape should define the fits function?

Reply:
yes! excellent point.